lib/spack/spack/build_environment.py

```diff
@@ -283,10 +283,15 @@ def set_build_environment_variables(pkg, env, dirty):
         except spack.spec.NoLibrariesError:
             tty.debug("No libraries found for {0}".format(dep.name))
 
-        for default_lib_dir in ['lib', 'lib64']:
-            default_lib_prefix = os.path.join(dep.prefix, default_lib_dir)
-            if os.path.isdir(default_lib_prefix):
-                dep_link_dirs.append(default_lib_prefix)
+            # Fallback to default lib/lib64 ONLY if the package failed to
+            # provide working libs(). Otherwise with complicated Intel packages
+            # we may end up including un-versioned lib folder
+            # /apps/intel/ComposerXE2017/lib instead of
+            # /apps/intel/ComposerXE2017/compilers_and_libraries_2018.3.222/...
+            for default_lib_dir in ['lib', 'lib64']:
```

Comment:
I don't know if there are any actual cases of this, but there may be packages which depend on the default behavior (adding `lib/`/`lib64/`) in addition to implementing their own `libs` method.

Generally speaking, adding `lib`/`lib64` wouldn't be necessary for the default implementation of `libs`, which starts searching from the root prefix:

```
        libs = fs.find_libraries(
            name, spec.prefix, shared=shared, recursive=True)
```

In that case, `lib/` would automatically be added if any libraries were found there.

This issue would come up only for packages with a `libs` implementation that starts searching from a sub-directory of `lib/` or `lib64/`. Do you see any packages like that?

If so, I think we'll also want to update the `LibraryList` object to allow adding directories explicitly (rather than implicitly as part of adding library files).

Reply:
I think the current default is to search `prefix.lib`, then `prefix.lib64`, then search `prefix` recursively. This is necessary for external installations where `/usr/` or `/` could take quite a long time to search recursively.

Reply:
That's true, although to be clear I don't see that addressing the concern I brought up in https://github.com/spack/spack/pull/10628#discussion_r475836464 (not to say you were asserting such, just want to make sure we are on the same page that there are are some remaining semantics to sort out).